### PR TITLE
Various changes to BIP9

### PR DESCRIPTION
**Pull request by sipa**
* Add start time
* Specify the state transitions less ambiguously.
* Add a state diagram.

### CODE REVIEW

**Comment by sipa** [comment]
@gmaxwell @rustyrussell @CodeShark Feel free to comment here

**Comment by sdaftuar** [comment]
@sipa I have a question about this sentence:

> Having more than 29 available bits for parallel soft forks does not add anything anyway, as the (nVersion >= 3) requirement already makes that impossible.

This statement seems incorrect: couldn't we just use 01 as the first two bits, which would satisfy all prior ISM rollouts while also leaving 30 bits available for parallel softfork deployment?  Am I missing something?

**Comment by sipa** [comment]
Nice catch. The sentence dates from before the 001 prefix was introduced.

**Comment by jtimon** [comment]
Yeah, that was very confusing. On the other hand, from what I understand there's still only 29 bits available since the first bit (aka hardfork bit) is still invalid and the other zero in the prefix is reserved for future use.

**Comment by jtimon** [comment]
Also, I insist the threshold is per-chain while the start time, bit and time out are per-deployment.
@sipa this is what you have implemented. I challenge you to implement a per-deployment threshold without completely breaking the warning check.

File: bip-0009.mediawiki [hunk 1]
```diff
@@ -18,136 +18,136 @@ This document specifies a proposed change to the semantics of the 'version' fiel
...
 
-'''Soft Fork Support'''
-Software which supports the change should begin by setting B in all blocks
-mined until it is resolved.
+# '''DEFINED''' is the first state that each soft fork starts out as. The genesis block is by definition in this state for each deployment.
+# '''STARTED''' for blocks past the starttime.
+# '''LOCKED_IN''' for one retarget period after the first retarget period with STARTED blocks of which at least threshold have the associated bit set in nVersion.
+# '''ACTIVE''' for all blocks after the LOCKED_IN retarget period.
+# '''FAILED''' for one retarget period past the timeout time, if LOCKED_IN was not reached.
+# '''ABANDONED''' for all blocks after the FAILED retarget period.
```
Comment (jtimon) [inline bip-0009.mediawiki]: why this additional state? isn't enough that it fails that it also has to be abandoned?

Comment (CodeShark) [inline bip-0009.mediawiki]: I think having a retargeting period where the bit cannot be reused is a good idea. Having an extra state can force this.

**Comment by CodeShark** [comment]
@sipa I gave it a once over and I really like what you did.

**Comment by FelixWeis** [comment]
really like the proposal. Maybe just make it a little more clear how a bit becomes available again once it reached one of the final stages.

**Pull request merged**